### PR TITLE
Add a --timeout option to `daml ledger` commands

### PR DESCRIPTION
**Pull request by cocreature**
At least in the compatibility tests, we see timeouts fairly regularly
in the tests for fetch-dar. If we see those timeouts in the tests, I
see no reason why our users should never see that especially on
actually distributed ledgers and hardcoded timeouts are a bad idea
anyway so now they are configurable.

I’ve also changed the timeout for the fetch tests (those are the only
once I’ve seen timeout) to 120s.

changelog_begin

- [DAML Assistant] The `daml ledger` commands now accept a `--timeout`
  option which can be used to change the timeout for API requests from
  the default of 30s.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
Note that this isn’t going to solve all flakiness in the compat tests since it clearly does not affect older versions. The only option I see there is to mark the tests exclusive or increase their resource allocation so less things run in parallel and timeouts are less likely to happen.

File: compatibility/bazel_tools/daml_ledger/Main.hs [hunk 1]
```diff
@@ -249,3 +283,16 @@ writeMinimalProject (SdkVersion sdkVersion) = do
...
+    ver >= SdkVersion
+        ( SemVer.initial
+        & SemVer.major .~ 1
+        & SemVer.minor .~ 1
+        & SemVer.patch .~ 1
+        )
+
+-- | TODO bump to the first snapshot after 1.4.0-snapshot.20200715.4733.0.d6e58626.
+supportsTimeout :: SdkVersion -> Bool
+supportsTimeout ver = ver >= SdkVersion SemVer.initial
```
Comment (garyverhaegen-da) [inline compatibility/bazel_tools/daml_ledger/Main.hs:298]: Can't you just say `>` and avoid having to remember this next week?

File: compatibility/bazel_tools/test_utils/DA/Test/Util.hs [hunk 2]
```diff
@@ -15,3 +17,5 @@ assertFileExists file = doesFileExist file >>= assertBool (file ++ " was expecte
 assertFileDoesNotExist :: FilePath -> IO ()
 assertFileDoesNotExist file = doesFileExist file >>= assertBool (file ++ " was expected to not exist, but does exist") . not
 
+assertInfixOf :: String -> String -> Assertion
+assertInfixOf needle haystack = assertBool ("Expected " <> show needle <> " in output but but got " <> show haystack) (needle `isInfixOf` haystack)
```
Comment (garyverhaegen-da) [inline compatibility/bazel_tools/test_utils/DA/Test/Util.hs:21]:
```suggestion
assertInfixOf needle haystack = assertBool ("Expected " <> show needle <> " in output but got " <> show haystack) (needle `isInfixOf` haystack)
```

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs [hunk 3]
```diff
@@ -337,6 +338,14 @@ commandParser = subparser $ fold
         <> metavar "TOKEN_PATH"
         <> help "Path to the token-file for ledger authorization"
 
+    timeoutOption :: Parser TimeoutSeconds
+    timeoutOption = option auto $ mconcat
+        [ long "timeout"
+        , metavar "TIMEOUT"
+        , value 30
+        , help "Timeout of gRPC operations in seconds. Defaults to 30s"
```
Comment (garyverhaegen-da) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs:346]:
```suggestion
        , help "Timeout of gRPC operations in seconds. Defaults to 30s. Must be > 0."
```

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs [hunk 4]
```diff
@@ -337,6 +338,14 @@ commandParser = subparser $ fold
         <> metavar "TOKEN_PATH"
         <> help "Path to the token-file for ledger authorization"
 
+    timeoutOption :: Parser TimeoutSeconds
+    timeoutOption = option auto $ mconcat
+        [ long "timeout"
+        , metavar "TIMEOUT"
```
Comment (garyverhaegen-da) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs:344]:
```suggestion
        , metavar "INT"
```

File: compatibility/bazel_tools/daml_ledger/Main.hs (hunk 1 above)
Comment (cocreature) [reply compatibility/bazel_tools/daml_ledger/Main.hs:301]: No 1.4.0 is not > 0.0.0

Comment (cocreature) [reply compatibility/bazel_tools/daml_ledger/Main.hs:301]: Oh you mean the current snapshot, yeah that works. I’ll change it.

Comment (garyverhaegen-da) [reply compatibility/bazel_tools/daml_ledger/Main.hs:301]: Well of course it isn't, I meant `> 1.4.0-snapshot.20200715.4733.0.d6e58626`.

**Comment by cocreature** [comment]
> Also, IIUC this is a server-side timeout. Is there/should there be a way for clients to also specify a timeout on their side?

My understanding was that gRPC timeouts are enforced on the client side but I might be wrong here. I’m not too keen on adding two kinds of timeouts. Both because adding any layer over the already very fragile gRPC library is a mess and because it’s a confusing UX.

> Is there a section of the documentation where this would be worth mentioning?

Added some docs to the best section I could find.

File: compatibility/bazel_tools/daml_ledger/Main.hs (hunk 1 above)
Comment (cocreature) [reply compatibility/bazel_tools/daml_ledger/Main.hs]: changed.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs (hunk 3 above)
Comment (garyverhaegen-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs:346]: The reason I think this is worth mentioning is there are timeout APIs out there where `-1` or `0` have a special "unlimited" meaning.

**Comment by dasormeter** [comment]
@moritzkiefer-da / @garyverhaegen-da strong 👍 on this one : we get these timeouts all the time especially during initial integration with distributed ledgers because they are slow for operations like package upload, especially.  Would be great to have this configurable for all the `daml ledger` operations

**Pull request merged**